Added remaining languagues to monaco editor, matched them with correct file extensions and editor modes

HNOONa-0 opened the pull request:
Added remaining languages to `util.js` and updated `index-light.html`options so that you can pick any of these languages. Now the languages of monaco editor matches that of the modal. Made sure to add correct file extensions and modes for each language.
Before (live version):
![l6](https://github.com/omegaup/omegaup/assets/59091536/7cd44d52-8d1d-42d0-b00e-55fafded8faf)
After:
![image](https://github.com/omegaup/omegaup/assets/59091536/71cef282-15af-464b-a631-1569ead8b46f)
![image](https://github.com/omegaup/omegaup/assets/59091536/42e3847b-cbbe-4f9b-bbc5-d5b83dbbfaba)

I wanted to refactor all attributes of a language like this:
```
// TO DO: later iterations, refactor language logic using this object.
// this object should carry all information needed about each language
export const languageInfo = {
  c: { extension: 'c', name: 'C11 (gcc 10.3)', modelMapping: 'cpp' },
  'c11-gcc': { extension: 'c', name: 'C11 (gcc 10.3)', modelMapping: 'cpp' },
  ....
  }
 ```
And then export it throughout the project, but I think that will take some time and debugging. I believe this pr fixes the issue for now.


# Description

Please include a description of the changes you have made. The text you put in this section will be used as the **commit message** when this PR is merged.

If you modify the **User Interface**, please include a screenshot or a video.

Fixes: #7483

# Comments

Add any comments for the reviewers (e.g., indicating the beginning of the revision, something where the reviewer needs to pay special attention, etc.). If  there are no extra comments this section can be deleted.

# Checklist:

- [x] The code follows the [coding guidelines](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) of omegaUp.
- [x] The tests were executed and all of them passed.

pabo99 reviewed (commented):
Thank you for fixing this.

Do you have any ideas on how to refactor it easily to obtain the same list of languages both in the `MonacoEditor` and in our `RunSubmit` popup?

If it's too complicated, we can leave a TODO and address it when we start working on the migration of the `EphemeralGrader` component.

HNOONa-0 commented:
@pabo99 For only two files its easy. My problem is its not only these two files, there are a lot of files that use language data locally, copy paste style. I wanted to have a single reference to this language object to solve this problem entirely.

but I will work on it.

HNOONa-0 commented:
Hey @pabo99, anything else would you like to cover? I could refactor more code around language logic in other files, but I think I should submit another issue for that.

pabo99 commented:
> Hey @pabo99, anything else would you like to cover? I could refactor more code around language logic in other files, but I think I should submit another issue for that.

I have some suggestions, but I need PR #7489 to be merged into the `main` branch. In that PR, I added the same language object that you created in the file `utils.js`. The difference is that the new one is a `TypeScript` file. It will help us to have better control over typing.

For now, I have a question: have you tried filling the options for the dropdowns dynamically? This way, all the language lists would be the same.

HNOONa-0 commented:
> > Hey @pabo99, anything else would you like to cover? I could refactor more code around language logic in other files, but I think I should submit another issue for that.
> 
> I have some suggestions, but I need PR #7489 to be merged into the `main` branch. In that PR, I added the same language object that you created in the file `utils.js`. The difference is that the new one is a `TypeScript` file. It will help us to have better control over typing.
> 
> For now, I have a question: have you tried filling the options for the dropdowns dynamically? This way, all the language lists would be the same.

You mean dropdown menu of the editor? that's part of the `index-light.html` file, I added the options (values to be selected) manually since it's a vanilla html file. If it was a vue component we could easily add languages dynamically as options.

We could instead handle it when `MonacoEditor` mounts. After mounting, fetch language input tag by id, update options of the input tag so using `supportedLanguages` object inside of `util.js`.

pabo99 commented:
Leave them as they are for now. I hope to migrate all those files to Vue + TypeScript soon to make our lives easier.

HNOONa-0 commented:
> Leave them as they are for now. I hope to migrate all those files to Vue + TypeScript soon to make our lives easier.

Understandable. I have one small request, can we run tests on this pr? I want to make sure that it passes all tests (except for the plagiarism test we talked about).

Also I plan on rewriting some parts of grader in ts, i will start by rewriting the vuex store thats inside of `ephemeral.js`, is that okay?

pabo99 commented:
Sure. Go ahead

HNOONa-0 commented:
Hello @pabo99, can we please start merging this pr too?

pabo99 commented on frontend/www/js/omegaup/components/arena/RunSubmitPopup.vue:
```diff
@@ -84,7 +84,9 @@ import T from '../../lang';
 import omegaup_Countdown from '../Countdown.vue';
 import omegaup_OverlayPopup from '../OverlayPopup.vue';
-
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import { supportedLanguages } from '../../grader/util';
```
Is there a way to not using this kind of comments?

If you are using it because is a javacript file and it doesn't have a declaration file, you can resolve as follows:

- Create the `util.ts` file with this constant and then you can import it here.
- Create the constant in this file and then you can export it to anywhere.

HNOONa-0 replied:
> Is there a way to not using this kind of comments?
> 
> If you are using it because is a javacript file and it doesn't have a declaration file, you can resolve as follows:
> 
> * Create the `util.ts` file with this constant and then you can import it here.
> * Create the constant in this file and then you can export it to anywhere.

I added `util.ts` file, initial type for the language `LanguageInfo` and deleted unnecessary objects from `util.js`. Added remaining languages to `index-light` in a separate commit.

HNOONa-0 replied:
I added `util.ts` file, initial type for the language `LanguageInfo` exported to the popup. We cannot delete `util.js` because  its being used at some components of the editor.

HNOONa-0 commented on frontend/www/js/omegaup/grader/util.js:
```diff
@@ -39,22 +39,28 @@ export function parseDuration(value) {
 
 export const languageMonacoModelMapping = {
-  cpp11: 'cpp',
+  c: 'cpp',
```
Done! Like you said, data can be derived from `supportedLanguages` constant

HNOONa-0 commented on frontend/www/grader/ephemeral/index-light.html:
```diff
@@ -32,17 +32,30 @@
 					<span class="fa fa-file-archive-o" title="Download" aria-hidden="true" id="download-label"></span>
 				</a>
 				<select class="form-control form-control-sm mr-sm-2" data-language-select id="language">
+					<option value="c">C11 (gcc 10.3)</option>
```
Done! Added as a separate commit.

pabo99 commented on frontend/www/js/omegaup/grader/util.js:
```diff
@@ -38,62 +38,170 @@ export function parseDuration(value) {
-  in: 'text',
-  out: 'text',
-  err: 'text',
+export const supportedLanguages = {
```
Is this constant neccesary here? Since you implemented in the `util.ts` file maybe it could be removed

HNOONa-0 replied:
Yes, `ephemeral.js` and `MonacoEditorComponent.vue` use it.

pabo99 replied:
Oh, you're right. I forgot we still need migrate `MonacoEditorComponent.vue` component to `TypeScript` as well

pabo99 reviewed (commented):
One last comment

pabo99 reviewed (approved):
LGTM
